pkg/query/api.go

```diff
@@ -0,0 +1,91 @@
+package query
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/improbable-eng/thanos/pkg/store/storepb"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/storage"
+)
+
+type Instant interface {
+	QueryInstant(ctx context.Context, qs string, ts time.Time, opts Options) (promql.Value, []error, error)
+}
+
+type Range interface {
+	QueryRange(ctx context.Context, qs string, start, end time.Time, interval time.Duration, opts Options) (promql.Value, []error, error)
+}
+
+type API struct {
+	engine *promql.Engine
+
+	replicaLabel string
+	storeAPI     storepb.StoreServer
+}
+
+func NewAPI(engine *promql.Engine, replicaLabel string, storeAPI storepb.StoreServer) *API {
```

Comment:
I think I will factor this out to just use PromQL natively in `api.go` and new middlewares.